server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -291,12 +274,11 @@ public void addTimeoutListener(@Nullable final TimeValue timeout, final TimeoutC
             threadPoolExecutor.execute(new SourcePrioritizedRunnable(Priority.HIGH, "_add_listener_") {
                 @Override
                 public void run() {
+                    final NotifyTimeout notifyTimeout = new NotifyTimeout(listener, timeout);
                     if (timeout != null) {
-                        NotifyTimeout notifyTimeout = new NotifyTimeout(listener, timeout);
                         notifyTimeout.cancellable = threadPool.schedule(notifyTimeout, timeout, ThreadPool.Names.GENERIC);
-                        onGoingTimeouts.add(notifyTimeout);
                     }
-                    timeoutClusterStateListeners.add(listener);
+                    timeoutClusterStateListeners.put(listener, notifyTimeout);
```

Comment:
The existing code seems to accept that `addTimeoutListener` is called multiple times for the same listener. It is not entirely clear to me that this cannot happen? If it does, a previous `notifyTImeout` will now go lost and not be cancelled.

We should at least add an assert here to verify, but maybe you can also do a bit of checking of whether there are legitimate cases where this could happen?

Reply:
I think tit's impossible for this to happen. Only `ClusterStateObserver` uses this method and it's using this code to make sure we only add a listener for it once:

```java
            final ObservingContext context = new ObservingContext(listener, statePredicate);
            if (!observingContext.compareAndSet(null, context)) {
                throw new ElasticsearchException("already waiting for a cluster state change");
            }
            clusterApplierService.addTimeoutListener(timeoutTimeLeftMS == null ?
                null : new TimeValue(timeoutTimeLeftMS), clusterStateListener);
```

-> I'm adding an assertion, I think that should be fine.